CIP-21: Governable LookbackWindow Smart Contract changes

andres-dg opened the pull request:
### Description

* Implementation of [CIP-21](https://github.com/celo-org/celo-proposals/pull/145) smart contract changes.
* celo-blockchain related PR: https://github.com/celo-org/celo-blockchain/pull/1136

Adds the `uptimeLookbackWindow` to the `BlockchainParameters` contract. This is only ever used on the `celo-blockchain` client to determine the size of the uptime lookback window which is used to compute the validator [uptimeScore](https://docs.celo.org/celo-codebase/protocol/proof-of-stake/epoch-rewards/validator-rewards#calculating-uptime-score)

An important invariant, is that `UptimeLookbackWindow` must remain constant during an epoch. Thus, changes to the parameter only take effect on the next epoch. There's an exception to this rule during migration/deployment of this change; there `uptimeLookbackWindow` should be initialized to chain's current value for the parameter which is part of the `ChainConfig` on the genesis block. 

To initialize the value:
  * On a new testnet: As part of the migrations, `BlockchainParamets.initialize()` will do the initialization
  * On an existing testnet: After upgrading the contract, first call to `setLookbackWindow()` will the value for the current epoch.

Since `BlockchainParameters` won't be reinitialized when this change is deployed, a patch was made to the `getUptimteLookbackWindow()` contract to return `12` (mainnet, baklava & alfajores value for lookbackWindow) in the case of the value not being initalized. This patch can be removed in the future, since it's only necessary until `setLoobackWindow()` is called and first epoch since deployment has passed

### Tested

* Smart Contract unit tests

adigiovanni97 committed:
CIP-21 Make lookback window governable

Changes:
 * Add new field & setter/getter to BlockchainParameters.sol
 * Add to contract migrations
 * Add to contractkit wrapper
 * Add contract version

nambrot commented on .circleci/config.yml:
```diff
@@ -31,7 +31,7 @@ e2e-defaults: &e2e-defaults
   docker:
     - image: celohq/circleci:geth1.9-2
   environment:
-    CELO_BLOCKCHAIN_BRANCH_TO_TEST: master
```
This probably should be back to master right?

mcortesi replied:
Not for the moment, since it's not merged there yet

nambrot replied:
I guess for this to merge into master, this needs to be master no?

nambrot commented on packages/protocol/contracts/governance/BlockchainParameters.sol:
```diff
@@ -75,6 +88,15 @@ contract BlockchainParameters is Ownable, Initializable {
+   * @notice Sets the uptime lookback window.
+   * @param window New window.
+   */
+  function setUptimeLookbackWindow(uint256 window) public onlyOwner {
```
You should check that the input is within reasonable bounds

prestwich replied:
what would reasonable bounds be for this parameter?

andres-dg replied:
The proposed limits, according to the [CIP](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0021.md), are [12 (1 minute) - 720 (1 hour)], I'm setting those for now.

Edit: This is assuming the block period is 5 seconds.

nambrot replied:
I assume a window of 0 or 1 day should be impossible?

mcortesi replied:
Added check as specified on https://github.com/celo-org/celo-proposals/pull/145/files

nambrot commented on packages/protocol/test/governance/network/blockchainparams.ts:
```diff
@@ -114,30 +116,58 @@ contract('BlockchainParameters', (accounts: string[]) => {
+      )
+    })
+  })
+
```
Add a check for malformed input

mcortesi committed:
Make lookbackWindow stable during epoch + testing

Changes the contract to delay modification of lookbackWindow value until
the next epoch.

Add tests and parameter range check

mcortesi committed:
Validate lookbackWindow is <= epochSize -2

mcortesi committed:
Add test for 1st and 2nd set during same epoch

mcortesi committed:
fix: using old method

prestwich commented on packages/protocol/contracts/governance/BlockchainParameters.sol:
```diff
@@ -75,6 +98,51 @@ contract BlockchainParameters is Ownable, Initializable {
+  function setUptimeLookbackWindow(uint256 window) public onlyOwner {
+    require(window >= 12 && window <= 720, "UptimeLookbackWindow must be within safe range");
+    require(
+      window <= getEpochSize() - 2,
```
edge case: if `epochSize` is 1 or 0 this will underflow

nambrot replied:
+1, use SafeMath

prestwich commented on packages/protocol/contracts/governance/BlockchainParameters.sol:
```diff
@@ -75,6 +98,51 @@ contract BlockchainParameters is Ownable, Initializable {
+      lookbackWindow = uptimeLookbackWindow.oldValue;
+    }
+
+    return lookbackWindow;
```
nit: unnecessary return statement

nambrot commented on packages/protocol/contracts/governance/BlockchainParameters.sol:
```diff
@@ -75,6 +98,51 @@ contract BlockchainParameters is Ownable, Initializable {
+      uptimeLookbackWindow.nextValueActivationEpoch = getEpochNumber();
+    } else {
+      // changes only take place on the next epoch
+      uptimeLookbackWindow.nextValueActivationEpoch = getEpochNumber() + 1;
```
SafeMath

nambrot commented on packages/protocol/contracts/governance/BlockchainParameters.sol:
```diff
@@ -75,6 +98,51 @@ contract BlockchainParameters is Ownable, Initializable {
+  function getUptimeLookbackWindow() public view returns (uint256 lookbackWindow) {
+    // epoch 0 is the genesis block, thus implies it has not been initialized
+    if (uptimeLookbackWindow.nextValueActivationEpoch == 0) {
+      // since on mainet,baklava,alfajores we don't reinitialize the contract
```
```suggestion
      // since on mainet,baklava,alfajores we haven't initialized the contract
```

nambrot commented on packages/protocol/test/governance/network/blockchainparams.ts:
```diff
@@ -114,30 +125,144 @@ contract('BlockchainParameters', (accounts: string[]) => {
+    it('if unset should set the variable for the current epoch', async () => {
+      await blockchainParameters.setUptimeLookbackWindow(24)
+      assert.equal((await blockchainParameters.getUptimeLookbackWindow()).toNumber(), 24)
+    })
```
new line in between tests

nambrot commented on packages/protocol/test/governance/network/blockchainparams.ts:
```diff
@@ -114,30 +125,144 @@ contract('BlockchainParameters', (accounts: string[]) => {
+      assert.equal((await blockchainParameters.getUptimeLookbackWindow()).toNumber(), 40)
+    })
+
+    it('if unset should set the variable for the current epoch', async () => {
```
I think this would benefit in readability if these pre-conditions are set with sub-describe's and beforeEach's

nambrot commented on packages/protocol/test/governance/network/blockchainparams.ts:
```diff
@@ -114,30 +125,144 @@ contract('BlockchainParameters', (accounts: string[]) => {
+      await blockchainParameters.setUptimeLookbackWindow(24)
+      assert.equal((await blockchainParameters.getUptimeLookbackWindow()).toNumber(), 24)
+    })
+    it('if unset, second set should set the variable for the next epoch', async () => {
```
Is this not the same test as on line 146?

mcortesi replied:
No, on this case we do 2 sets within the same epoch, and one is the initial one

nambrot commented on packages/protocol/test/governance/network/blockchainparams.ts:
```diff
@@ -114,30 +125,144 @@ contract('BlockchainParameters', (accounts: string[]) => {
+    it('should fail when using value lower than safe minimum', async () => {
+      await assertRevert(
+        blockchainParameters.setUptimeLookbackWindow(11, {
+          from: accounts[1],
```
This will revert just because it's not the owner

mcortesi replied:
dumb :roll_eyes: 
Thanks

nambrot commented on packages/protocol/test/governance/network/blockchainparams.ts:
```diff
@@ -114,30 +125,144 @@ contract('BlockchainParameters', (accounts: string[]) => {
+    it('should fail when using value greater than safe maximum', async () => {
+      await assertRevert(
+        blockchainParameters.setUptimeLookbackWindow(721, {
+          from: accounts[1],
```
ditto

nambrot commented on packages/protocol/test/governance/network/blockchainparams.ts:
```diff
@@ -114,30 +125,144 @@ contract('BlockchainParameters', (accounts: string[]) => {
+      )
+    })
+
+    it('should fail when using value greater than epochSize-2', async () => {
```
```suggestion
    it('should fail when using value greater than epochSize - 2', async () => {
```

mcortesi committed:
Address pr comments

nambrot commented on packages/protocol/test/governance/network/blockchainparams.ts:
```diff
@@ -141,94 +141,95 @@ contract('BlockchainParameters', (accounts: string[]) => {
+    describe('when not initialized', () => {
+      it('should set the variable for the current epoch', async () => {
+        await blockchainParameters.setUptimeLookbackWindow(initialValue)
+        assert.equal(
```
You'd probably technically want to make this assertion before setting in case initialValue === 12

mcortesi replied:
There's already a tests for that:
```ts
    it('should return 12 if not initalized', async () => {
      assert.equal((await blockchainParameters.getUptimeLookbackWindow()).toNumber(), 12)
    })
```

nambrot replied:
What I'm saying is that if `initialValue` is 12, then a faulty setUptimeLookbackWindow (i.e. set in the next epoch) would still make this test pass,right?

oneeman replied:
Do you mean before setting to have the same assert using `.notEqual` to ensure the before and after values can't coincidentally be equal?

mcortesi committed:
fixed

oneeman commented on packages/protocol/contracts/governance/BlockchainParameters.sol:
```diff
@@ -75,6 +101,49 @@ contract BlockchainParameters is Ownable, Initializable {
+    if (uptimeLookbackWindow.nextValueActivationEpoch == 0) {
+      // on initialize we set the value for the current epoch which configurer
+      // must make it so it's SAME as geth's chainConfig value
+      uptimeLookbackWindow.nextValueActivationEpoch = getEpochNumber();
```
Why is this necessary?  Suppose we always used `getEpochNumber().add(1)`, would that be problematic?

mcortesi replied:
For the case where you deploy the contract mid-epoch. 

The change we just did on celo-blockchain helps with that... but only if the contract is already there when donut is activated.

mcortesi committed:
Fix uptimeComputation on tests, and validator order test

mcortesi committed:
Add log msgs to avoid CI stopping by inactivity

mcortesi committed:
Configure blocktime on genesis, not cli

mcortesi committed:
Simplify lookbackWindow logic

mcortesi committed:
Fix contract & tests

oneeman commented on packages/protocol/contracts/governance/BlockchainParameters.sol:
```diff
@@ -109,35 +109,31 @@ contract BlockchainParameters is Ownable, Initializable, UsingPrecompiles {
-    uptimeLookbackWindow.oldValue = getUptimeLookbackWindow();
+    uptimeLookbackWindow.oldValue = _getUptimeLookbackWindow();
+
+    // changes only take place on the next epoch
```
This should also be made clear in the function's @notice above.

oneeman commented on packages/protocol/contracts/governance/BlockchainParameters.sol:
```diff
@@ -109,35 +109,31 @@ contract BlockchainParameters is Ownable, Initializable, UsingPrecompiles {
-      return 12;
-    } else if (getEpochNumber() >= uptimeLookbackWindow.nextValueActivationEpoch) {
+    lookbackWindow = _getUptimeLookbackWindow();
+    require(lookbackwindow != 0, "UptimeLookbackWindow is not initialized");
```
I'd avoid the word 'initialized' here since it can be initialized but not yet in effect.  If we want to be super-helpful we could give a different error in the case where it hasn't been initialized and the case where it has been initialized but we haven't yet passed to the next epoch.  Otherwise, I would phrase the error as "UptimeLookbackWindow not set for current epoch" or something like that.

oneeman commented on packages/protocol/contracts/governance/BlockchainParameters.sol:
```diff
@@ -109,35 +109,31 @@ contract BlockchainParameters is Ownable, Initializable, UsingPrecompiles {
   }
 
   /**
    * @notice Gets the uptime lookback window.
```
Call out that this reverts if there is no lookback window in effect for the current epoch (I don't know if there is a special way to do that or if it's just mentioning that in the @notice)

oneeman commented on packages/protocol/contracts/governance/BlockchainParameters.sol:
```diff
@@ -109,35 +109,31 @@ contract BlockchainParameters is Ownable, Initializable, UsingPrecompiles {
+  }
+
+  /**
+   * @notice Gets the uptime lookback window.
```
Call out that this returns 0 if there is no lookback window value in effect for the current epoch

mcortesi committed:
fix tests

mcortesi committed:
Merge remote-tracking branch 'origin/master' into andres-dg/loopbackwindow-refactor

mcortesi committed:
lint fix

mcortesi committed:
Merge branch 'master' into andres-dg/loopbackwindow-refactor
